File: wear/src/main/java/technology/mainthread/apps/gatekeeper/messenger/WearableMessenger.java
package technology.mainthread.apps.gatekeeper.messenger;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.PendingResult;
import com.google.android.gms.wearable.MessageApi;
import com.google.android.gms.wearable.Node;
import com.google.android.gms.wearable.NodeApi;
import com.google.android.gms.wearable.Wearable;

import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import technology.mainthread.apps.gatekeeper.common.SharedValues;
import technology.mainthread.apps.gatekeeper.injector.WearAppClient;

public class WearableMessenger {

    private GoogleApiClient googleApiClient;

    @Inject
    public WearableMessenger(@WearAppClient GoogleApiClient googleApiClient) {
        this.googleApiClient = googleApiClient;
    }

    /**
     * Send a wearable message using the MessageApi
     *
     * @param path message identifier for the receiver
     * @param data payload
     */
    public boolean sendMessage(String path, byte[] data) {
        return sendMessageInternal(path, data);
    }

    private boolean sendMessageInternal(String path, byte[] data) {
        boolean result = false;
        if (isConnected()) {
            String nodeId = pickBestNodeId(getNodes());
            PendingResult<MessageApi.SendMessageResult> pendingResult =
                    Wearable.MessageApi.sendMessage(googleApiClient, nodeId, path, data);

            MessageApi.SendMessageResult messageResult = pendingResult.await();
            result = messageResult.getStatus().isSuccess();
        }
        disconnect();

        return result;
    }

    private boolean isConnected() {
        if (googleApiClient.isConnected()) {
            return true;
        } else {
            ConnectionResult connectionResult = googleApiClient.blockingConnect(
                    SharedValues.CONNECTION_TIME_OUT_MS, TimeUnit.MILLISECONDS);
            return connectionResult.isSuccess();
        }
    }

    private void disconnect() {
        if (googleApiClient.isConnected()) {
            googleApiClient.disconnect();
        }
    }

    private List<Node> getNodes() {
        NodeApi.GetConnectedNodesResult nodes = Wearable.NodeApi.getConnectedNodes(googleApiClient).await();
        return nodes.getNodes();
    }

    private String pickBestNodeId(List<Node> nodes) {
        String bestNodeId = null;
        // Find a nearby node or pick one arbitrarily
        for (Node node : nodes) {
            if (node.isNearby()) {
                return node.getId();
            }
            bestNodeId = node.getId();
        }
        return bestNodeId;
    }

}

File: wear/src/main/java/technology/mainthread/apps/gatekeeper/messenger/WearableMessenger.kt
package technology.mainthread.apps.gatekeeper.messenger

import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.wearable.Node
import com.google.android.gms.wearable.Wearable
import technology.mainthread.apps.gatekeeper.common.CONNECTION_TIME_OUT_MS
import technology.mainthread.apps.gatekeeper.injector.WearAppClient
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class WearableMessenger @Inject
constructor(@param:WearAppClient private val googleApiClient: GoogleApiClient) {

    /**
     * Send a wearable message using the MessageApi

     * @param path message identifier for the receiver
     * *
     * @param data payload
     */
    fun sendMessage(path: String, data: ByteArray): Boolean {
        return sendMessageInternal(path, data)
    }

    private fun sendMessageInternal(path: String, data: ByteArray): Boolean {
        var result = false
        if (isConnected) {
            val nodeId = pickBestNodeId(nodes)
            val pendingResult = Wearable.MessageApi.sendMessage(googleApiClient, nodeId, path, data)

            val messageResult = pendingResult.await()
            result = messageResult.status.isSuccess
        }
        disconnect()

        return result
    }

    private val isConnected: Boolean
        get() {
            if (googleApiClient.isConnected) {
                return true
            } else {
                val connectionResult = googleApiClient.blockingConnect(
                        CONNECTION_TIME_OUT_MS, TimeUnit.MILLISECONDS)
                return connectionResult.isSuccess
            }
        }

    private fun disconnect() {
        if (googleApiClient.isConnected) {
            googleApiClient.disconnect()
        }
    }

    private val nodes: List<Node>
        get() {
            val nodes = Wearable.NodeApi.getConnectedNodes(googleApiClient).await()
            return nodes.nodes
        }

    private fun pickBestNodeId(nodes: List<Node>): String? {
        var bestNodeId: String? = null
        // Find a nearby node or pick one arbitrarily
        for (node in nodes) {
            if (node.isNearby) {
                return node.id
            }
            bestNodeId = node.id
        }
        return bestNodeId
    }

}
